proposals/0000-namespaces.md

```diff
@@ -0,0 +1,796 @@
+This proposal also introduces using another access qualifier for
+namespaces to differentiate them from modules (`:`).
+
+## Motivation
+
+The proposal will enable us to write succinct code and have better control of the exports.
+
+**The goal** is to 
+1. add first-class namespaces to Haskell, 
+2. with semantic and syntactic changes so small that it can be enabled by default, 
+3. which enables incremental adoption of more advanced features.
+
+**Why?** currently, due to the lack of namespaces, if you have two constructors, types, functions, patterns,
+or type classes named the same, you have one of three options:
+1.  Use the module system and move the different entities to different modules and import them qualified.
+    This have several down-sides:
+    
+    a) it promotes many small files which can be hard to get an overlook at;
+    
+    a) modules cannot have cyclic dependencies, eliminating some cases;
+    
+    b) it is hard to teach and communicate as some example are spread over multiple files; and
+    
+    c) this is not an option when using Haskell as a scripting language, as single files are often a requirement.
+
+2.  Use type classes to capture similarities. This leads to developers creating type classes without any meaning, purely
+    to capture similar named functions. This solution also breaks down fast if the similar named items have
+    different arity or arguments.
+
+3.  Renaming things. This is the most common solution. Pre- or Suffixing namespace identifies to solutions; 
+    e.g. `prettyExpr`, `prettyType`, `prettyVar`, `MkT`, `runState`, `applyEndo`, and so on. 
+    Besides (subjectivly) being akward to use and read, it has multiple drawbacks.
+    
+    a) It prompotes unqualified import of modules, e.g., `State.runState` seems redundant.
+       This is turn will pull more items into scope making it harder to present good completion
+       options for the user of an IDE.
+    
+    b) With suffix notation is hard to discover the abilities of a types using an IDE, e.g., 
+       typing `pretty` will yield `{prettyExpr, prettyType, prettyVar}`, but typing 
+       `Expr:` could yield `{Expr:pretty, Expr:evaluate, Expr:normalize, ...}`.
+    
+    c) The trick does not work for operators.
+
+The rest of the motivation is a list of examples mirroring [#283](https://github.com/ghc-proposals/ghc-proposals/pull/283)
+
+### Example 1: Compact code
+
+If we want associate the same named function with two data structures, we'll have 
+to create two files, create a type class, or prefix the names of the function.
+
+With namespaces you can write this:
+
+```haskell
+module Vec where
+
+namespace V2 where
+  data V2 = Mk { x :: Int, y :: Int }
+  fromV2 :: V2 -> V2
+
+namespace V3 where
+  data V3 = Mk { x :: Int, y :: Int, z :: Int }
+  fromV2 :: V2 -> Int -> V3
```

Comment:
Shouldn't this be

```suggestion
  fromV2 :: V2:V2 -> Int -> V3
```

I may be misunderstanding the proposal, but I don't see why `V2` is in scope here otherwise.

Reply:
Yes, you are correct. That example have been changed a lot. It's fixed now.